### PR TITLE
adds support for screen readers to active menu item

### PR DESCRIPTION
**Pull request by N6REJ**
Pull Request for Issue # .

### Summary of Changes
adds `aria-current="page"` to active menu item.

### Testing Instructions
select a menu item to make it active.
view source for active menu item.
should look like "actual result" image.
apply patch
`aria-current="page"` should be before the `>` in the anchor for the "active" menu item.
as shown in the "expected result" image.

Now select or create as needed a alias menu item
repeat above steps.

### Expected result

image

### Actual result
image

### Documentation Changes Required

None

### CODE REVIEW

File: modules/mod_menu/tmpl/default.php [hunk 1]
```diff
@@ -31,6 +31,7 @@
 	if ($item->id == $active_id || ($item->type === 'alias' && $item->params->get('aliasoptions') == $active_id))
 	{
 		$class .= ' current';
+		$currnt = TRUE;
```
Comment (SharkyKZ) [inline modules/mod_menu/tmpl/default.php]:
```suggestion
		$current = true;
```

File: modules/mod_menu/tmpl/default.php [hunk 2]
```diff
@@ -95,7 +96,7 @@
 	// The next item is on the same level.
 	else
 	{
-		echo '</li>';
+		echo $current ? '<span class="sr-only">(current)</span></a></li>' : '</li>';
```
Comment (SharkyKZ) [inline modules/mod_menu/tmpl/default.php]: Text should be translatable.

Comment (SharkyKZ) [reply modules/mod_menu/tmpl/default.php]: There is no `sr-only` class in Protostar.

File: modules/mod_menu/tmpl/default.php (hunk 1 above)
Comment (SharkyKZ) [reply modules/mod_menu/tmpl/default.php]: Variable has to be overwritten always, not just when condition is met.

File: modules/mod_menu/tmpl/default.php (hunk 2 above)
Comment (N6REJ) [reply modules/mod_menu/tmpl/default.php]: @SharkyKZ 
sr-only is specifically to trigger screen readers is it not?
@brianteeman seems to be the expert on accessibility so perhaps I'm doing it all wrong.

File: modules/mod_menu/tmpl/default.php (hunk 1 above)
Comment (N6REJ) [reply modules/mod_menu/tmpl/default.php]: stupid typo 📦

File: modules/mod_menu/tmpl/default.php (hunk 2 above)
Comment (SharkyKZ) [reply modules/mod_menu/tmpl/default.php]: The equivalent in Protostar is `element-invisible`.

Comment (C-Lodder) [reply modules/mod_menu/tmpl/default.php]: @N6REJ `.sr-only` is a custom Bootstrap class, not a native browser class ;)

**Comment by brianteeman** [comment]
The right idea but completely the wrong implementation ;)

You just need to use an aria attribute called aria-current on the active link - in this case it would be aria-current="page"

See https://tink.uk/using-the-aria-current-attribute/

note this needs to be done in j4 admin menu as well if you want to submit a pr for that

**Comment by N6REJ** [comment]
> aria-current="page"

ok, as soon as I figure out why GIT is telling me to flack off I'll try to fix it.. ty.

**Comment by N6REJ** [comment]
@brianteeman would it be applied to the list item or the anchor?

**Comment by brianteeman** [comment]
Exactly as it is written in the page I linked to

File: modules/mod_menu/tmpl/default.php (hunk 2 above)
Comment (laoneo) [reply modules/mod_menu/tmpl/default.php]: From a quick check this class is not available in BS 2.3.

Comment (brianteeman) [reply modules/mod_menu/tmpl/default.php]: It's being rewritten the correct way without a class ;)

**Comment by N6REJ** [comment]
~~Can't seem to figure out what creates the menu item when the menu type is "single article"~~
**SOLVED**

**Comment by N6REJ** [comment]
@brianteeman ready for testing 👯‍♂

**Comment by brianteeman** [comment]
Why are you excluding setting aria-current when its the home?

**Comment by N6REJ** [comment]
what install?  I don't see that behavior
image

**Comment by brianteeman** [comment]
sorry - I misread the diff - i think - will test later

**Comment by N6REJ** [comment]
turns out that when the menu item is an alias ( like top bar home in protostar ) although it gets the "active & current" classes it's ID doesn't match the active id value because that is to its referenced menu item.  SO, there's another id that references its parent so you have to test against that one too ergo the last commit

**Review by N6REJ — COMMENTED** [review]
why is it saying I've changed all those lines in default.php which I CLEARLY haven't per this screen shot?
image
I'm especially referring to 26-29 and 36-42
my change other then formatting is 34

**Comment by SharkyKZ** [comment]
It is because you changed the formatting. You can see changes without whitespace changes here https://github.com/joomla/joomla-cms/pull/26261/files?w=1.

**Comment by N6REJ** [comment]
> It is because you changed the formatting. You can see changes without whitespace changes here https://github.com/joomla/joomla-cms/pull/26261/files?w=1.

whew!! ty @SharkyKZ my chest is literally pounding cause I thought I somehow totally messed things up and was freaking out trying to figure out how the hell I did that!
image
This is what freaked me out.
How did you find the link for w/o formatting?

**Comment by SharkyKZ** [comment]
When viewing changed files, click the gear icon and select `Hide whitespace changes`.

**Comment by brianteeman** [comment]
Pr for joomla 4 admin menu done #26326

File: libraries/src/HTML/HTMLHelper.php [hunk 3]
```diff
@@ -329,6 +329,8 @@ protected static function includeRelativeFiles($folder, $file, $relative, $detec
 			return array($file);
 		}
 
+		$file = \JPath::clean($file, '/');
```
Comment (brianteeman) [inline libraries/src/HTML/HTMLHelper.php]: why is this here? messed up branches again?

Comment (N6REJ) [reply libraries/src/HTML/HTMLHelper.php]: yep!  fixed.

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on c5b3f23afb73c8aac8f7d7a39ee81ca2829f5c85<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by brianteeman** [comment]
Adding aria-current is perfectly ok
Removing or changing classes is not. This will potentially break any site that is using those classes for styling<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by SharkyKZ** [comment]
>Removing or changing classes is not.

@brianteeman What are you talking about?

**Comment by brianteeman** [comment]
I have not tested this item.<br /><br />Reset my test - the way github displayed the diff it made it look like classes were changed<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

File: modules/mod_menu/tmpl/default_component.php [hunk 4]
```diff
@@ -57,4 +57,9 @@
 	$attributes['onclick'] = "window.open(this.href, 'targetWindow', '" . $options . "'); return false;";
 }
 
+if ($active_id == $item->id || $active_id == $item->params['aliasoptions'])
+{
+	$attributes['aria-current'] = "page";
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_component.php]:
```suggestion
	$attributes['aria-current'] = 'page';
```

File: modules/mod_menu/tmpl/default_url.php [hunk 5]
```diff
@@ -63,4 +63,9 @@
 	$attributes['onclick'] = "window.open(this.href, 'targetWindow', '" . $options . "'); return false;";
 }
 
+if ($active_id == $item->id || $active_id == $item->params['aliasoptions'])
+{
+	$attributes['aria-current'] = "page";
```
Comment (Quy) [inline modules/mod_menu/tmpl/default_url.php]:
```suggestion
	$attributes['aria-current'] = 'page';
```

**Comment by N6REJ** [comment]
made changes suggested by @Quy

**Comment by Renuka-S** [comment]
I have tested this item :white_check_mark: successfully on b27290347988cd4093d37fc261e2cf6c887a4589<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by Renuka-S** [comment]
I have tested this item :white_check_mark: successfully on b27290347988cd4093d37fc261e2cf6c887a4589<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by jduerscheid** [comment]
I have tested this item :white_check_mark: successfully on b27290347988cd4093d37fc261e2cf6c887a4589<br /><br />It works<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by HLeithner** [comment]
There are some things wrong in this PR.

in default_component.php you try to check if the active_id is in aliasoptions, thats not possible because aliasoptions is only used if the type is "alias". So you can remove this.

You use the same if in default_url.php which will is also wrong. In this case $item->id can't be $active_id reason for this is that in this case the type is "url" and url can't be active. Also the second part of this if statement is wrong because you try to access $item->params as array but it's a JRegistry for this you have to access it with the ->get method.

The other thing is the attribute it self, Joomla's menu system is really weak for the current page. it uses anything that has the correct Itemid as current. Which basically doesn't need to be true for example for blog posts in categories. For example if you read a article in a "news" category the menu item is highlighted but it doesn't link to the current article instead it links to the "news" category.

The j4 implementation is more correct using the complete url to set this attribute.

**Comment by brianteeman** [comment]
> The other thing is the attribute it self, Joomla's menu system is really weak for the current page. it uses anything that has the correct Itemid as current. Which basically doesn't need to be true for example for blog posts in categories. For example if you read a article in a "news" category the menu item is highlighted but it doesn't link to the current article instead it links to the "news" category.

You are confusing things here. The purpose of this PR is to inform users of assistive technology with information that tells them exactly which menu item is the active one - ie the menu item that was clicked last. Your example seems to be talking about a canonical link which is not the same thing at all.

**Comment by HLeithner** [comment]
Your PR for J4 #26326 adds aria-current="page" only if the complete url is the same. Which would be "the current page". But if you add aria-current="page" to <a href="/news">news</a> on a page with the url "/news/5-article" doesn't sound correct to me.

Why should it be different in j4 then in j3? ymmv

**Comment by brianteeman** [comment]
> Your PR for J4 #26326 adds aria-current="page" only if the complete url is the same.

No thats not correct. It does what I believe the code here does and adds it to the active menu item

> Why should it be different in j4 then in j3? ymmv

I dont believe (I've not checked the code here) it is just that the implementation is different

At the end of the day the objective of this aria attribute is to provide an accessible alternative to adding css to a menu link to indicate it is the currently selected menu link

**Comment by brianteeman** [comment]
NOTE my J4 pr is for the admin - this is for the frontend - where the sub-items that dont have menu links (which is what I guess your example would be) have  the main menu hidden. So we're probably both getting a little confused here as we're comparing apples to lemons I'm not even sure if I am convinced that this should be in core

**Comment by HLeithner** [comment]
@alikon I can't remove RTC, also I think the implementation as I wrote above isn't correct. Also I'm not sure if this attribute is correct at all in this case.

**Comment by Quy** [comment]
Removed RTC.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26261">issues.joomla.org/tracker/joomla-cms/26261</a>.</sub>

**Comment by N6REJ** [comment]
I give up.  It's always a fight to do anything.
Someone else can take the heat.
